File: frontend/src/types/device.ts
/**
 * Type definitions for Device data structures
 */

export type DeviceStatus =
  | "unknown"
  | "backup_success"
  | "backup_failed"
  | "backup_in_progress"
  | "backup_no_changes"

export interface Device {
  device_name: string
  ip_address: string
  vendor: string
  group: string
  ssh_profile: string
  protocol: string
  port: number
  enabled: boolean
  status: DeviceStatus
  last_backup: string | null
  last_backup_success: string | null
  last_error?: string | null
  schedule?: string | null
}

export interface DeviceGroup {
  group: string
  count: number
}

File: backend/app/models/device.py
"""Device-related Pydantic models."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, IPvAnyAddress


class DeviceStatus(str, Enum):
    """Device backup status."""

    UNKNOWN = "unknown"
    BACKUP_SUCCESS = "backup_success"
    BACKUP_FAILED = "backup_failed"
    BACKUP_IN_PROGRESS = "backup_in_progress"
    BACKUP_NO_CHANGES = "backup_no_changes"


class DeviceBase(BaseModel):
    """Device configuration (static fields from source)."""

    device_name: str = Field(..., min_length=1, max_length=255)
    ip_address: IPvAnyAddress
    vendor: str = Field(..., min_length=1, max_length=255)
    group: str = Field(..., min_length=1, max_length=255)
    ssh_profile: str = Field(..., min_length=1, max_length=255)
    protocol: str = Field(..., min_length=1, max_length=16)
    port: int = Field(..., ge=1, le=65535)
    enabled: bool = True


class DeviceFull(DeviceBase):
    """Device config + backup/status information."""

    status: DeviceStatus = DeviceStatus.UNKNOWN
    last_backup: Optional[datetime] = None
    last_backup_success: Optional[datetime] = None
    last_error: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
